### ISSUE TITLE
Pageant is too trusting

### ISSUE THREAD

**Issue by ferricoxide** [NoMoreFood/putty-cac#79]
**Summary**

If Windows has cached a CAC-certificate, Pageant will allow you to load and forward that CAPI certificate without verifying that the CAC is still inserted or that you have the right to use it (no PIN prompt).

**Steps to reproduce**

On a system with a secondary card-reader:

1. Insert CAC (or alt-token) into secondary card-reader
1. Execute an action that causes the card to be read and PIN-verified
1. Remove the CAC (or alt-token) from the secondary card-reader
1. Launch Pageant
1. Click on the `Add CAPI cert` button
1. On the certificate-selection pop-up, click on the "more" button
1. Select the removed certificate from the certificate list
1. Click the `Ok` button
1. Pageant inserts the certificate into its key-list without requesting PIN or even verifying that the CAC (or alt-token) is still inserted in the reader 

**Background**

Discovered this on accident after having launched a VDI session from within another operator's on-site, physical workstation. The next day, when I launched a new VDI session as part of my telework activities, I'd autopiloted my way through my morning-routine of firing up Pageant and loading the first cert that showed up after hitting the `Add CAPI cert` button. Didn't notice I'd selected the wrong token (prior day's physical workstation-owner's). Went to login to our development-environment's bastion host and was prompted for my user's password rather than being immediately-presented with a prompt. After supplying my password, I completed the login process. 

The bastion, itself isn't CAC-enabled. However, our git service is. I do remote-development on an EC2 that I have to reach from the bastion-host and use the key-forwarding feature to enable git-over-SSH with my CAPI token. Since I'm already forwarding the token, I usually set up my bastion's `${HOME}/.ssh/authorized_keys` to include my CAPI to save myself having to enter my  policy-compliant password every time I need to open a new session or re-establish the session because the network has dropped, yet again.

Typically, I end up being prompted for my password due to one of two reasons:
* the bastion-owners have executed one of their periodic bastion-replacements and my `${HOME}` has gotten nuked in the process (the usual reason)
* PuTTY has failed to forward the key (usually only happens when the VDI service's session-template has been updated resulting in my VDI profile having gotten reset, killing my PuTTY registry-settings along with it)

So, I went to check what the cause was:
* Checked my `${HOME}/.ssh` directory (present) and its `authorized_keys` file (also present and contained my CAPI token's public-key)
* Checked my SSH auth-agent session and found that "my" key _had_ been forwarded

This caused me to go back to Pageant and right-scroll the list-keys view, where I noticed that the CAPI token was showing the wrong distinguished name. This in turn caused me to dig further and notice that Windows had stored the prior day's workstation-owner's key and that Pageant had simply taken it as an article of faith that I was allowed to use all of the cached keys.

**Comment by NoMoreFood** [comment]
I'm not sure I completely understand what your reported issue is.

Depending on policy settings, certificates will be automatically imported into your personal Windows certificate store when a token is inserted into the system.  You can remove orphaned ones by typing 'Manage user certificates' into the Start Menu.  If you don't want Pageant to automatically load these certificates then do not use the 'Autoload Certs' feature.  If Pageant were to force-prompt for PIN during the Pageant automatic load process (to confirm you have access to the private key), it would be extremely inconvenient for users that have multiple legitimate tokens that they do not use on every session.

**Comment by ferricoxide** [comment]
If you're talking about a PuTTY-CAC feature, I don't believe that I'm using the "autoload certs" feature. I'm talking about, when you're using Pageant  (say at the beginning of the day) to load a given key (or set of keys) for the purposes of key-forwarding and you're explicitly-selecting which key to load.

At no point in the CAC-load – or continuing CAC-usage – process does Pageant ever prompt for a PIN. Worse, if Windows has unexpectedly cached a another user's client-cert, Pageant is more than happy to ingest that certificate without any validation of the right to do so (and, similarly, more than happy to forward that key).

This is at odds with the observed behavior of any other CAC-enabled tooling I've been using over the years. As near as I can tool, that other tooling implement an initial-validation combined with a validation-TTL. Typically, this means that on initial access (say when first logging into a CAC-protected application or web-page) the user is prompted for their PIN. Similarly, after "X" hours since the last, successful PIN-prompt, the application or web page will re-request the user validate their PIN.

**Comment by chuckmilam** [comment]
I'm not seeing this behavior.   

My workflow:

1. At the beginning of the day, I start pageant, right-click and select "Add CAPI Cert."  I am not prompted for a PIN at this time. 
2. When I attempt to open a session, I am prompted for my PIN.  This only works if the smardcard with the correct CAPI certificate I chose to add is inserted in a card reader.
3. For some VERY short period of time, my certificate is cached.  I can usually run a command, pull my card, then run it again.  This is good for less than a minute, more like a few seconds.  After that, I will be re-prompted for my PIN, which will only work if the required smartcard is inserted.
4. If I remove my smartcard (step away from my desk, etc.) I will be prompted for my PIN again when I return and attempt to run a command.

My setup:

1. Pageant does NOT have any of these possibly relevant options selected: 

- Autoload Certs
- Remember Certs
- Force PIN Caching
- Cert Auth Prompting

2.  I do have the ActiveClient CAC/smartcard middleware installed, so we're not using the native Windows smartcard interface.

Obviously, this is not a VDI setup, and I'm the only real user on this particular physical workstation.  So, I'm wondering if this may be a VDI issue?  Improperly shared/isolated registry entries, perhaps?

**Comment by ferricoxide** [comment]
@chuckmilam 

Dunno. I was completely surprised when I tripped over it. Given that, when I'd accessed my VDI session it had been from another user's thick-client and I _only_ did a PIN-login to the VDI with _my_ CAC, it was wholly unexpected that the thick-client's owner's certificate got cached into my VDI session. At no point during the use of my VDI session did I explicitly-reference his CAC. It was even more surprising that his CAC cert was cached until explicitly-purged, given that I'd wholly logged out of my VDI (it would have been automatically logged out overnight, regardless) rather than disconnecting and that I logged into a new session the next day. 

Frankly, when it comes to our VDI's deployment, it's more surprising when registry and other settings persist than when they don't: it's so normal for browser bookmarks, PuTTY session-profiles, etc. to disappear that I've long since habituated to exporting the configs any time I add or modify a configuration-element.

I'll have to check on the Pageant settings. PuTTY comes to our VDI by an app-streaming service. I've never bothered to look into modifying any of the settings because many such streaming apps lose any setting mods between launches (GIMP being a prime example).

Likely a moot point, any way: we're soon moving to having to SSH through an authentication-proxy service that very much _does_ PIN-prompt each time you initiate a new connection through the service. Even knowing this was very much an edge-case scenario, I wanted to bring it up "just in case" (especially given that how I even noticed was that Pageant had allowed the not-owned-by-me and not-present CAC-cert to be forwarded).
